File: code/scenes/src/main/java/org/fbme/scenes/controllers/scene/SceneStateManager.java
package org.fbme.scenes.controllers.scene;

import jetbrains.mps.openapi.editor.EditorComponent;
import jetbrains.mps.openapi.editor.cells.EditorCell;
import jetbrains.mps.openapi.editor.extensions.EditorExtension;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.mps.openapi.model.SNodeReference;

import java.util.HashMap;
import java.util.Map;

public class SceneStateManager {

    private static final Map<EditorComponent, SceneStateManager> INSTANCES = new HashMap<>();

    public static final EditorExtension EDITOR_EXTENSION = new EditorExtension() {

        public boolean isApplicable(@NotNull EditorComponent ec) {
            return true;
        }

        public void install(@NotNull EditorComponent ec) {
            INSTANCES.put(ec, new SceneStateManager());
        }

        public void uninstall(@NotNull EditorComponent ec) {
            SceneStateManager stateManager = INSTANCES.remove(ec);
            stateManager.myStates.clear();
        }
    };

    private final Map<StateId, Object> myStates = new HashMap<>();

    public SceneStateManager() {
    }

    public static SceneStateManager getInstance(@NotNull EditorComponent ec) {
        return INSTANCES.get(ec);
    }

    public <T> void storeState(EditorCell cell, SceneStateKey<T> key, T state) {
        myStates.put(new StateId(cell.getSNode().getReference(), cell.getCellId(), key), state);
    }

    public <T> T loadState(EditorCell cell, SceneStateKey<T> key) {
        //noinspection unchecked
        return (T) myStates.get(new StateId(cell.getSNode().getReference(), cell.getCellId(), key));
    }

    private static class StateId {
        private final SNodeReference myNodeRef;
//        private final String myCellId;
        private final SceneStateKey<?> myKey;

        private StateId(SNodeReference nodeRef, String cellId, SceneStateKey<?> key) {
            myNodeRef = nodeRef;
//            myCellId = cellId;
            myKey = key;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (!(o instanceof StateId)) {
                return false;
            }

            StateId that = (StateId) o;
//            return myNodeRef.equals(that.myNodeRef) && myCellId.equals(that.myCellId) && myKey.equals(that.myKey);
            return myNodeRef.equals(that.myNodeRef) && myKey.equals(that.myKey);
        }

        @Override
        public int hashCode() {
            int result = 0;
            result = 31 * result + myNodeRef.hashCode();
//            result = 31 * result + myCellId.hashCode();
            result = 31 * result + myKey.hashCode();
            return result;
        }
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/scene/SceneStateManager.kt
package org.fbme.scenes.controllers.scene

import jetbrains.mps.openapi.editor.EditorComponent
import jetbrains.mps.openapi.editor.cells.EditorCell
import jetbrains.mps.openapi.editor.extensions.EditorExtension
import org.jetbrains.mps.openapi.model.SNodeReference

class SceneStateManager {
    private val states: MutableMap<StateId, Any?> = HashMap()

    fun <T> storeState(cell: EditorCell, key: SceneStateKey<T>, state: T) {
        states[StateId(cell.sNode.reference, cell.cellId, key)] = state
    }

    fun <T> loadState(cell: EditorCell, key: SceneStateKey<T>): T? {
        return states[StateId(cell.sNode.reference, cell.cellId, key)] as T?
    }

    private class StateId(
        private val nodeRef: SNodeReference,
        private val cellId: String?,
        private val key: SceneStateKey<*>
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) {
                return true
            }
            if (other !is StateId) {
                return false
            }
//            return nodeRef.equals(other.nodeRef) && cellId.equals(other.cellId) && key.equals(other.key);
            return nodeRef == other.nodeRef && key == other.key
        }

        override fun hashCode(): Int {
            var result = 0
            result = 31 * result + nodeRef.hashCode()
//            result = 31 * result + cellId.hashCode();
            result = 31 * result + key.hashCode()
            return result
        }
    }

    companion object {
        private val INSTANCES: MutableMap<EditorComponent, SceneStateManager> = HashMap()
        val EDITOR_EXTENSION: EditorExtension = object : EditorExtension {
            override fun isApplicable(ec: EditorComponent): Boolean {
                return true
            }

            override fun install(ec: EditorComponent) {
                INSTANCES[ec] = SceneStateManager()
            }

            override fun uninstall(ec: EditorComponent) {
                val stateManager = INSTANCES.remove(ec)
                stateManager!!.states.clear()
            }
        }

        fun getInstance(ec: EditorComponent): SceneStateManager? {
            return INSTANCES[ec]
        }
    }
}
